net: if: Add functions to check any pending TX or RX packets

jukkar opened the pull request:
These functions can be used for example by network power management
to check if the network interface can be suspended or not.
If there are network packets in transmit queue, then the network
interface cannot be suspended yet.
The corresponding RX functions have limited use but were created
for symmetry.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: if: Add function to check any pending TX packets

This function can be used for example by network power management
to check if the network interface can be suspended or not.
If there are network packets in transmit queue, then the network
interface cannot be suspended yet.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

tbursztyka commented on include/net/net_if.h:
```diff
@@ -469,6 +469,10 @@ struct net_if {
 	struct net_if_config config;
+
+	/** Keep track of packets pending in traffic queues */
+	int tx_pending[NET_TC_TX_COUNT];
```
Do we care to track per-queue? 
Also, I fail to see any other usage than with CONFIG_NETWORK_POWER_MANAGEMENT  so we could avoid growing the size of the structure when this option is unset. What do you think ?

jukkar replied:
Yes, we could go first with global counter. Also the RX tracking is not needed (I just put it there for symmetry). Noticed also a bug where the code did not check whether the net-pkt was inserted properly to the workqueue. Will send a new version shortly.

jukkar replied:
> CONFIG_NETWORK_POWER_MANAGEMENT

BTW, should we call this CONFIG_NET_POWER_MANAGEMENT instead, to be similar like other network related config options?

tbursztyka replied:
yep

tbursztyka replied:
heh, actually it is already like that. Hard to focus these days...

tbursztyka commented on include/net/net_if.h:
```diff
@@ -469,6 +469,10 @@ struct net_if {
+
+	/** Keep track of packets pending in traffic queues */
+	int tx_pending[NET_TC_TX_COUNT];
+	int rx_pending[NET_TC_RX_COUNT];
```
Do we need to track these? Currently, net pm only requires to know about tx queue.

jukkar committed:
net: tc: Return information whether TX pkt was queued

We need to know whether the net_pkt was successfully placed
to transmit queue. It is possible in TX side, that the net_pkt
is already in TX queue when for example TCP packet is
re-transmitted, in which case the queue submit will fail.
This cannot happen in RX side as there are no timers involved.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
@tbursztyka can you test this with your PR and if no issues we can merge this first and then your network PM PR #23067

jukkar commented:
Ouch, sanitychecker checks the Kconfig symbol and it is missing now. I think only option is that you take these two commits and merge them to your PR and we close this one.

tbursztyka commented:
@jukkar I am closing this PR, I merged the 2 commits in #23067
While testing your commits I always got "device busy", so looking closely to the tx_pending attribute, I found that it was going below 0. I finally fixed your first commit, the work queue was indeed fired right after k_work_submit_to_queue(), so further k_work_pending() was returning false.
